drivers/dma/dma_stm32_v1.c

```diff
@@ -0,0 +1,301 @@
+	LL_DMA_ClearFlag_TE7,
+};
+
+static void (*func_clear_dme[])(DMA_TypeDef *DMAx) = {
+	LL_DMA_ClearFlag_DME0,
+	LL_DMA_ClearFlag_DME1,
+	LL_DMA_ClearFlag_DME2,
+	LL_DMA_ClearFlag_DME3,
+	LL_DMA_ClearFlag_DME4,
+	LL_DMA_ClearFlag_DME5,
+	LL_DMA_ClearFlag_DME6,
+	LL_DMA_ClearFlag_DME7,
+};
+
+static void (*func_clear_fe[])(DMA_TypeDef *DMAx) = {
+	LL_DMA_ClearFlag_FE0,
+	LL_DMA_ClearFlag_FE1,
+	LL_DMA_ClearFlag_FE2,
+	LL_DMA_ClearFlag_FE3,
+	LL_DMA_ClearFlag_FE4,
+	LL_DMA_ClearFlag_FE5,
+	LL_DMA_ClearFlag_FE6,
+	LL_DMA_ClearFlag_FE7,
+};
+
+static u32_t (*func_is_active_te[])(DMA_TypeDef *DMAx) = {
+	LL_DMA_IsActiveFlag_TE0,
+	LL_DMA_IsActiveFlag_TE1,
+	LL_DMA_IsActiveFlag_TE2,
+	LL_DMA_IsActiveFlag_TE3,
+	LL_DMA_IsActiveFlag_TE4,
+	LL_DMA_IsActiveFlag_TE5,
+	LL_DMA_IsActiveFlag_TE6,
+	LL_DMA_IsActiveFlag_TE7,
+};
+
+static u32_t (*func_is_active_dme[])(DMA_TypeDef *DMAx) = {
+	LL_DMA_IsActiveFlag_DME0,
+	LL_DMA_IsActiveFlag_DME1,
+	LL_DMA_IsActiveFlag_DME2,
+	LL_DMA_IsActiveFlag_DME3,
+	LL_DMA_IsActiveFlag_DME4,
+	LL_DMA_IsActiveFlag_DME5,
+	LL_DMA_IsActiveFlag_DME6,
+	LL_DMA_IsActiveFlag_DME7,
+};
+
+static u32_t (*func_is_active_fe[])(DMA_TypeDef *DMAx) = {
+	LL_DMA_IsActiveFlag_FE0,
+	LL_DMA_IsActiveFlag_FE1,
+	LL_DMA_IsActiveFlag_FE2,
+	LL_DMA_IsActiveFlag_FE3,
+	LL_DMA_IsActiveFlag_FE4,
+	LL_DMA_IsActiveFlag_FE5,
+	LL_DMA_IsActiveFlag_FE6,
+	LL_DMA_IsActiveFlag_FE7,
+};
+
+void stm32_dma_dump_stream_irq(DMA_TypeDef *dma, u32_t id)
+{
+	LOG_INF("tc: %d, ht: %d, te: %d, dme: %d, fe: %d",
+		func_is_active_tc[id](dma),
+		func_is_active_ht[id](dma),
+		func_is_active_te[id](dma),
+		func_is_active_dme[id](dma),
+		func_is_active_fe[id](dma));
+}
+
+void stm32_dma_clear_stream_irq(DMA_TypeDef *dma, u32_t id)
+{
+	func_clear_te[id](dma);
+	func_clear_dme[id](dma);
+	func_clear_fe[id](dma);
+}
+
+bool stm32_dma_is_irq_happened(DMA_TypeDef *dma, u32_t id)
+{
+	if (func_is_active_fe[id](dma)) {
+		return true;
+	}
+
+	return false;
+}
+
+bool stm32_dma_is_unexpected_irq_happened(DMA_TypeDef *dma, u32_t id)
+{
+	if (func_is_active_fe[id](dma)) {
+		LOG_ERR("FiFo error.");
+		stm32_dma_dump_stream_irq(dma, id);
+		stm32_dma_clear_stream_irq(dma, id);
+
+		return true;
+	}
+
+	return false;
+}
+
+void stm32_dma_enable_stream(DMA_TypeDef *dma, u32_t id)
+{
+	LL_DMA_EnableStream(dma, table_stream[id]);
+}
+
+int stm32_dma_disable_stream(DMA_TypeDef *dma, u32_t id)
+{
+
+	if (!LL_DMA_IsEnabledStream(dma, table_stream[id])) {
+		return 0;
+	}
+	LL_DMA_DisableStream(dma, table_stream[id]);
+
+	return -EAGAIN;
+}
+
+static void stm32_dma_config_direct_mode(DMA_TypeDef *dma, u32_t id, bool on)
+{
+	if (on) {
+		LL_DMA_EnableIT_DME(dma, table_stream[id]);
+		LL_DMA_DisableIT_FE(dma, table_stream[id]);
+		LL_DMA_DisableFifoMode(dma, table_stream[id]);
+	} else {
+		LL_DMA_DisableIT_DME(dma, table_stream[id]);
+		LL_DMA_EnableIT_FE(dma, table_stream[id]);
+		LL_DMA_EnableFifoMode(dma, table_stream[id]);
+	}
+}
+
+#define BURST_SIZE(dev, dev_short)					\
```

Comment:
Not quite convinced about this trick.
It saves few lines (macro in only called once, so we save only one "duplication" ) but at reading comfort expense..
Can you revert to usual code?

Reply:
Sure